tests/test_utils.py

```diff
@@ -950,3 +951,33 @@ def target():
     thread.start()
     thread.join()
     assert (main_thread.ident, main_thread.name) == results.get(timeout=1)
+
+
+@pytest.mark.parametrize(
+    ("input_timestamp", "expected_output"),
+    (
+        ("2021-01-01T12:00:00Z", "2021-01-01T12:00:00.000000Z"),  # UTC time
+        (
+            "2021-01-01T12:00:00+02:00",
+            "2021-01-01T10:00:00.000000Z",
+        ),  # Non-UTC time
+        (
+            "2021-01-01T12:00:00-07:00",
+            "2021-01-01T19:00:00.000000Z",
+        ),  # Another non-UTC time
+    ),
+)
+def test_format_timestamp(input_timestamp, expected_output):
+    datetime_object = datetime.fromisoformat(input_timestamp)
+    formatted = format_timestamp(datetime_object)
+
+    assert formatted == expected_output
+
+
+def test_format_timestamp_naive():
+    datetime_object = datetime.fromisoformat("2021-01-01T12:00:00")
+    timestamp_regex = r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{6}Z"
+
+    # Ensure that some timestamp is returned, without error. We currently treat these as local time, but this is an
+    # implementation detail which we should not assert here.
+    assert re.fullmatch(timestamp_regex, format_timestamp(datetime_object))
```

Comment:
Let's keep it simple and just check against the actual output we expect, no need to future-proof like this. If we check against the actual timestamp we can also add this as another case to the `parametrize` of `test_format_timestamp`.

Reply:
But this will depend on the local timezone of wherever we are running the tests. We cannot just have a single expected output

Reply:
We would basically need to reimplement `format_timestamp` in the unit test and assert against the output from there. Or, we would need some way to mock the system's timezone (not sure how difficult that would be, but probably some complication involved)

Reply:
Ok I incorrectly assumed naive timestamps don't have any implicit timezone so casting to UTC would just add the timezone. Just goes to show the general "timezones suck" lesson hasn't really stuck with me properly yet. 😆 

I guess this is fine then, even if it's a bit weird.